libs/groupSubscriptionManager.js

```diff
@@ -0,0 +1,160 @@
+var async = require('async');
+var uuid = require('uuid');
+var moment = require('moment');
+var request = require('request');
+
+var db, queue, amazonPayment, done, habitrpgUsers, jobStartDate, habitGroups;
+
+var plan = {
+  price: 3,
+  quantity: 3,
+};
+var pageLimit = 10;
+
+var SUBSCRIPTION_CANCEL_URL = 'https://habitica.com/amazon/subscribe/cancel';
+
+function scheduleNextQueue()
+{
+  queue.create('amazonGroupPlanPayments')
+    .priority('critical')
+    .delay(jobStartDate.add({hours: 1}).toDate() - new Date())
+    .attempts(5)
+    .save(function(err){
+      return err ? done(err) : done();
+    });
+}
+
+function cancelSubscription(group)
+{
+  habitrpgUsers.findOne({ _id: group.leader }, { castIds: false })
+    .then(function (user) {
+      request({
+        url: SUBSCRIPTION_CANCEL_URL + '?groupId=' + group._id,
+        method: 'GET',
+        qs: {
+          noRedirect: 'true',
+          _id: user._id,
+          apiToken: user.apiToken
+        }
+      }, function(error, response, body) {
+        console.log('error cancelling', error, body);
+      });
+    });
+}
+
+function chargeGroup (group, callback)
+{
+  var price = plan.price * (plan.quantity + group.memberCount - 1);
+
+  amazonPayment.authorizeOnBillingAgreement({
+    AmazonBillingAgreementId: group.purchased.plan.customerId,
+    AuthorizationReferenceId: uuid.v4().substring(0, 32),
+    AuthorizationAmount: {
+      CurrencyCode: 'USD',
+      Amount: price,
+    },
+    SellerAuthorizationNote: 'Habitica Subscription Payment',
+    TransactionTimeout: 0,
+    CaptureNow: true,
+    SellerNote: 'Habitica Subscription Payment',
+    SellerOrderAttributes: {
+      SellerOrderId: uuid.v4(),
+      StoreName: 'Habitica'
+    }
+  })
+  .then(function(response) {
+    // TODO should expire only in case of failed payment
+    // otherwise retry
+    if (response.AuthorizationDetails.AuthorizationStatus.State === 'Declined') {
+      cancelSubscription(group);
```

Comment:
what about transforming cancelSubscription in a function that returns a promise and returning it instead of just firing it without a callback?